File: Algorithm_66_Plus One.cpp
#include <iostream>
#include <vector>
using namespace std;
vector<int> plusOne(vector<int>& digits)
{
	int len=digits.size(),carry=0;
	digits[len-1]+=1;
	for (int i=len-1;i>=0;i--)
	{
		digits[i]+=carry;
		if (digits[i]==10)
		{
			digits[i]=0;
			carry=1;
		}
		else carry=0;
	}
	if (carry==1) digits.insert(digits.begin(),1);
	return digits;
}
int main()
{
	vector<int> num;
	num.push_back(0);
	//num.push_back(9);
	num=plusOne(num);
	for (int i=0;i<num.size();i++) cout<<num[i];
	return 0;
}

File: Algorithm_66_Plus One.java
// 对digits从后向前遍历，判断当前digits[i]是否为9。若不为9，则+1后不会产生进位，直接digits[i]++，然后返回digits
// 若为9，则把当前位置为0，i--，继续循环
// 若for正常结束，说明始终在产生进位，此时要返回的结果数组res长度比digits多1，其res[0]为1而其他位全部为0

class Solution {
    public int[] plusOne(int[] digits) {
        if (digits == null || digits.length == 0) {
            return new int[] {1};
        }
        for (int i = digits.length - 1; i >= 0; i--) {
            if (digits[i] != 9) {
                digits[i]++;
                return digits;
            } else {
                digits[i] = 0;
            }
        }
        int[] res = new int[digits.length + 1];
        res[0] = 1;
        return res;
    }
}